[Serve] [Doc] Copyedit and refactor doc structure for Managing Deployments (#26968)
ray-project/ray: doc/source/serve/configuring-serve-deployments.md

You can check the configuration using `cv2.getNumThreads()` and `cv2.getNumberOfCPUs()`.
:::

(managing-deployments-handling-dependencies)=

## Handling Dependencies

Ray Serve supports serving deployments with different (possibly conflicting)
Python dependencies.  For example, you can simultaneously serve one deployment
that uses legacy Tensorflow 1 and another that uses Tensorflow 2.

This is supported on Mac OS and Linux using Ray's {ref}`runtime-environments` feature.
As with all other Ray actor options, pass the runtime environment in via `ray_actor_options` in
your deployment.  Be sure to first run `pip install "ray[default]"` to ensure the
Runtime Environments feature is installed.

Example:

```{literalinclude} ../../../python/ray/serve/examples/doc/conda_env.py
```

:::{tip}

> Avoid dynamically installing packages that install from source: these can be slow and

Reviewer: It is not very clear what dynamically installing packages means here. From the conda_env.py example, does it count as dynamically installing packages ? :)

Author: Ah good point, let me improve this

Author: Maybe let's merge this and I'll make the improvement in a followup edit pass.  This PR is just to reorganize the doc structure

Author: Addressed it in this PR, it should be clearer now.

You can check the configuration using `cv2.getNumThreads()` and `cv2.getNumberOfCPUs()`.
:::

(managing-deployments-handling-dependencies)=

## Handling Dependencies

Ray Serve supports serving deployments with different (possibly conflicting)
Python dependencies.  For example, you can simultaneously serve one deployment
that uses legacy Tensorflow 1 and another that uses Tensorflow 2.

This is supported on Mac OS and Linux using Ray's {ref}`runtime-environments` feature;
see :ref:`Runtime Environments<runtime-environments>` for details.

As with all other Ray actor options, pass the runtime environment in via `ray_actor_options` in
your deployment.  Be sure to first run `pip install "ray[default]"` to ensure the
Runtime Environments feature is installed.

Example:

```{literalinclude} ../../../python/ray/serve/examples/doc/conda_env.py
```

:::{tip}
Certain `pip` packages build from source when installed via `pip install`; these can

> use up all resources while installing, leading to problems with the Ray cluster.  Consider

Reviewer 2: ```suggestion
use up too many resources while installing, leading to problems with the Ray cluster.  Consider
```

precompiling such packages in a private repository or Docker image.
:::

> The dependencies required in the deployment may be different than

Reviewer 2: ```suggestion
The dependencies required in the deployment may be different from
```

the dependencies installed in the driver program (the one running Serve API

> calls). In this case, you should use a delayed import within the class to avoid

Reviewer 2: ```suggestion
calls). In this case, you should use a delayed import within the deployed class to avoid
```
